### ISSUE TITLE
Consider removing the dependency on java.util.Optional

### ISSUE THREAD

**Issue by condesales** [carlanton/m3u8-parser#36]
`java.util.Optional` was only introduced on Android at API level 24 (7.0)

This can be restrictive for a lot of Android projects out there. The most common supported version is currently API level 21 (5.0)

There are multiple different approaches on how to replace this dependency, but I reckon the best would probably be implementing an `Optional` Class in the same structure as the java.util one.

something like this would probably be enough:

```java
import java.util.Objects;

public class Optional<T> {

    private T value;

    private Optional() {
        this.value = null;
    }

    private Optional(T value) {
        this.value = Objects.requireNonNull(value);
    }

    public static<T> Optional<T> empty() {
        return new Optional<>();
    }

    public static<T> Optional<T> of(T value) {
        return new Optional<>(value);
    }

    public interface Action<T> {
        void apply(T value);
    }

    public void ifPresent(Action<T> action) {
        if (value != null) {
            action.apply(value);
        }
    }
}
```

**Comment by carlanton** [comment]
Hi Felipe,

I think you have to give me a little more context for such a big change :-) How and why are you using this library on Android?
I'm not really interested to break the current API, but maybe it's possible to generate java 7 compatible code somehow?

Cheers,
Anton

**Comment by condesales** [comment]
Hi Anton, thanks for the quick response 👋 

> How and why are you using this library on Android?

Currently, I'm working on a project that supports video content via m3u8 URLs.
So far that's been no need to parse the m3u8 file myself as Exoplayer2 and Cast player were doing it for me just fine.

Now we're working on supporting track selection - more specifically subtitles.
Exoplayer2 support for tracking selection works (although not great) but for Cast player the information is non-existent on the Sender side.

I decided then to use the library to parse the manifest myself and get all information I required, instead of relying on players doing that for me. Probably a better approach as players might parse them differently and we want to give users a consistent UX.
That's when I came across this project.

> maybe it's possible to generate java 7 compatible code somehow

There is a possibility to add Java 8 support in the Android project by doing some setups, but in the same way you wanted to avoid breaking the current library API I wanted to avoid going down that path as it adds a lot of overhead to the project's compilation time 😅 

Thus why I created the issue with "**Consider** removing the dependency". Not sure to what extent the Java 7-8 features are used in the library, you're probably in a much better position to evaluate that, but if not extensively there are simple ways to remove it to make the library play ball with more environments.
